### ISSUE TITLE
[Feature]: Enhance plugin-swc

### ISSUE THREAD

**Issue by JSerFeng** [web-infra-dev/modern.js#4214]
## Function type config support

Some users need more detailed control of SWC, and there are many many configurations in SWC, currently `builder-plugin-swc` only exposes some  frequently used options, like the `react` and `env` fields of SWC configurations.

But some users may need some more config that we do not notice, like for mobx users, they need to set `useDefineforClass` option, I've never noticed this before, and I cannot guarantee that I can take all the configs into account in the future either, so it's necessary for users to have full control of SWC.

## Overrides config

Some users wanted that, enable a specific plugin only in file foo, and enable other plugins in file bar. Or if SWC minimizer has bugs in some files, we can disable minimize only in those files. It's similar to [babel's overrides](https://babeljs.io/docs/options#overrides)

### What does the proposed API look like?

We can support function type config, and pass some useful utilities.

```ts
function(swcConfig, { presetReact, presetEnv }) {
  swcConfig.jsc.transform.useDefineForClassFields = true;
  presetEnv({ coreJs: '2' })
}
```

And overrides like:

Object config:
```ts
{
  overrides: [ { include: ['./bug-file.ts'], jsMinify: false }  ]
}
```

Function config:
```ts
function(config, { addOverride }) {
  # helper utility
  addOverride('./bug-file.ts', { jsMinify: false })

  # or modify directly
  config.overrides.push({ include: [/node_modules\/foo/], {  jsc: { transform: { useDefineForClassFields: true } } } })
}
```

**Comment by chenjiahan** [comment]
Can we implement this in the framework level?

EdenX framework provides a [tools.swc](https://edenx.bytedance.net/en/configure/app/tools/swc.html) config, we can enhance it and support function type, just like `tools.webpack` and `tools.babel`.

**Comment by JSerFeng** [comment]
> Can we implement this in the framework level?

Yeah, that is what I thought, all the examples above, is for `tools.swc`, both for edenx and modern.js, not direct builder plugin

**Comment by chenjiahan** [comment]
Do we really need a `addOverride` util? 

It seems we are using different swc configs for different files. If users configure it, it may be difficult to migrate to Rspack in the future.

**Comment by JSerFeng** [comment]
> Do we really need a `addOverride` util?
> 
> It seems we are using different swc configs for different files. If users configure it, it may be difficult to migrate to Rspack in the future.

Overrides if just for framework level, under the hood it will just become different module.rules, like:

```
rules: [ { test: /foo.js/, loader: 'swc-plugins/loader', options: optionForFoo }, { test: /bar.js/, loader: 'swc-plugins/loader', options: optionForBar } ]
```

**Comment by chenjiahan** [comment]
> Overrides is just for framework level, under the hood it will just become different module.rules

Get, but we do not provide this feature for Babel, so I'm wondering if we need to provide it for SWC. Is there some real business needs here?

**Comment by chenjiahan** [comment]
Get, it makes sense.
